Enhance SearchableFinder to find searchable classes outside of app/

mfullbrook opened the pull request:
Enhance SearchableFinder to find searchable classes outside of `app/` directory.

This change enables scout-extended to work with non-typical Laravel projects that have models stored in directories outside of `app/`.  For example a project that uses the [laravel-modules](https://github.com/nWidart/laravel-modules) library.

* `SearchableFinder` uses config (if present) in `config/scout.php`.  The existing behaviour has not changed.  I'm unsure if this is the best approach as it requires the end user to add config that is not present in the template `scout.php`.
* Tests modified to test the discovery of models located in the `laravel/modules` directory.  All tests passing.

nunomaduro commented on composer.json:
```diff
@@ -39,7 +39,8 @@
             "Tests\\": "tests/",
-            "App\\": "tests/laravel/app"
+            "App\\": "tests/laravel/app",
+            "Modules\\": "tests/laravel/modules"
```
Can't we detect the `project_scan_dirs` and `project_namespaces` from this value?

mfullbrook replied:
Oh yeah, that seems obvious now!  

For the benefit of the tests we would need to fetch mappings from `autoload` and `autoload-dev`.  Are you aware of an API exposed by composer to read the active project's composer file?  When running the tests in this package the path to composer.json will be different to when vendored into a laravel project.

mfullbrook replied:
So i have a solution...
```
$composer = Composer\Factory::create(new Composer\IO\NullIO(), null,null);
$activeJsonPath = $composer->getConfig()->getConfigSource()->getName();
// $activeJsonPath == '/Users/mark/code/scout-extended/composer.json'
```

mfullbrook replied:
I've refactored by interrogating Composer - this enhancement is now zero configuration!

nunomaduro commented on composer.json:
```diff
@@ -17,6 +17,7 @@
         "php": "^7.1.3",
         "ext-json": "*",
         "algolia/algoliasearch-client-php": "^2.0",
+        "composer/composer": "^1.8",
```
Sorry, buddy. You should wait me to answer before coding. You don't accept new dependencies because this.

Laravel internally have a function get composer's contents. Use that one.

mfullbrook replied:
No worries - i was keen!

Out of interest, for a package maintainer, why is adding the dependency a bad idea when Laravel also has the same dependency?

`Illuminate\Support\Composer` is not an option as it is a facade to calling composer scripts. The composer path is not exposed as public method.   

Instead we are now using `base_path('composer.json')` and calling `json_decode`.  This is actually better than using `Composer\Composer` because we are definitely loading the laravel project's composer.json, not the scout-extended's.

nunomaduro replied:
> Out of interest, for a package maintainer, why is adding the dependency a bad idea when Laravel also has the same dependency?

I didn't know this. Anyway, seems way clean now.

nunomaduro commented on src/Helpers/SearchableFinder.php:
```diff
@@ -35,6 +36,16 @@ final class SearchableFinder
+    /**
+     * @var array
+     */
+    private $projectNamespaces;
```
I would prefer that you remove this.

nunomaduro commented on src/Helpers/SearchableFinder.php:
```diff
@@ -35,6 +36,16 @@ final class SearchableFinder
+    /**
+     * @var array
+     */
+    private $projectSources;
```
I would prefer that you remove this.

nunomaduro commented on src/Helpers/SearchableFinder.php:
```diff
@@ -71,10 +82,10 @@ public function fromCommand(Command $command): array
     public function find(): array
     {
-        $appNamespace = $this->app->getNamespace();
+        $this->inferProjectSourcePaths();
```
Can you get a `$sources` and a `$namespaces` from this method?

nunomaduro commented on src/Helpers/SearchableFinder.php:
```diff
@@ -105,4 +120,25 @@ private function getProjectClasses(): array
+     *
+     * @return void
+     */
+    private function inferProjectSourcePaths(): void
```
Return `$sources` and a `$namespaces` from this method instead of setting it into attributes of the class.

nunomaduro commented:
@mfullbrook Do you think you are able to work on this buddy?

mfullbrook commented:
Hi @nunomaduro, so sorry to have neglected this for a ridiculous amount of time🙈.

Anyway, I've refactored and fixed the broken tests. But, I have ten identical errors that, for me, are also present on master.

```
9) Tests\Features\TransformersTest::testTransformMethod
Algolia\AlgoliaSearch\Exceptions\UnreachableException: Impossible to connect, please check your Algolia Application Id.
```

I stepped through the trace and can see that the factory methods in the tests are triggering the dispatch of an `UpdateJob`.

Master is passing on travis, but master is failing on my local.  My branch is getting the same errors on Travis.  Any thoughts or pointers would be really appreciated.

jnbn commented:
Any updates on this? @mfullbrook @nunomaduro

mfullbrook commented:
Hi @DevinCodes, I've rebased my changed on top of master.  The tests all pass on my local machine when using my own Algolia Application.  Would you be able to give guidance as to why the tests are all failing on CircleCI with `Algolia\AlgoliaSearch\Exceptions\BadRequestException: Invalid Application-ID or API key`. Would really appreciate it. 🙏

DevinCodes commented:
Hey @mfullbrook ! Happy new year, and sorry for the delay in reply.

There's currently an issue with the CI for forks, so I had to rerun your tests manually: they are all green 🎉 
 I'll give the PR a final review today, I think we're good to go after that! I'll keep you posted.

DevinCodes reviewed (changes requested):
It looks good to me overall, thank you for your work on this!

Could you please rebase the PR on the `master` branch one last time? This ensures your PR is compatible with the latest release too, which added support for PHP8. Once that's done, I'll re-run the tests, and if all is green I think it's good to go.

mfullbrook commented:
@DevinCodes thanks! I'm using this branch in my own app and it's working great.  I've rebased, please re-run the tests when you can.

DevinCodes reviewed (approved):
Thank you for your work on this @mfullbrook ! I'll release a new version with this feature tomorrow 🙂

DevinCodes commented:
This is now available in [`v1.12.0`](https://github.com/algolia/scout-extended/releases/tag/v1.12.0)! 🎉
